examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlClient.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2016, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.examples.manualflowcontrol;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class ManualFlowControlClient {
+    public static void main(String[] args) throws InterruptedException {
+        final ExecutorService pool = Executors.newCachedThreadPool();
+        final Object done = new Object();
+
+        // Create a channel and a stub
+        ManagedChannel channel = ManagedChannelBuilder
+                .forAddress("localhost", 50051)
+                .usePlaintext(true)
+                .build();
+        GreeterGrpc.GreeterStub stub = GreeterGrpc.newStub(channel);
+
+        // When using manual flow-control and back-pressure on the client, the ClientResponseObserver handles both
+        // request and response streams.
+        ClientResponseObserver<HelloRequest, HelloReply> clientResponseObserver =
+                new ClientResponseObserver<HelloRequest, HelloReply>() {
+
+            ClientCallStreamObserver<HelloRequest> requestStream;
+
+            @Override
+            public void beforeStart(final ClientCallStreamObserver<HelloRequest> requestStream) {
+                this.requestStream = requestStream;
+                // Set up manual flow control for the response stream. It feels backwards to configure the response
+                // stream's flow control using the request stream's observer, but this is the way it is.
+                requestStream.disableAutoInboundFlowControl();
+
+                // Set up a back-pressure-aware producer for the request stream. The onReadyHandler will be invoked
+                // when the consuming side has enough buffer space to receive more messages.
+                //
+                // Note: the onReadyHandler is invoked by gRPC's internal thread pool. You can't block in this in
```

Comment:
I introduced the external thread pool on the server side specifically because all the callbacks are serialized. The `onReadyHandler` was waiting for work and was blocking `onNext` from accepting that work.

The client side needed an external thread because the `onReadyHandler` would have to run to completion before any server responses could be processed. This prevented true bi-directional streaming from happening. All the requests would be sent, _then_ all the responses would be consumed.